enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,444 @@
+
+#### Timeline & Phases
+
+| Phase                                                                          | Target       |   
+|--------------------------------------------------------------------------------|--------------|
+| Phase 1 - Out-Of-Tree SWAP with WASP TP                                        | mid 2024     |
+| Phase 2 - WASP GA with swap-based evictions                                    | end 2024     |
+| Phase 3 - Transition to swap operator. Limited to CNV-only users               | mid-end 2025 |
+| Phase 4 - Limits removed. All Openshift users allowed to use the swap operator | TBD          |
+
+* **Phase 1 (Done)** - OpenShift Virtualization will provide an out-of-tree
+  solution (WASP) to enable higher workload density and swap-based eviction mechanisms.
+* **Phase 2 (Done)** - WASP will include swap-based eviction mechanisms.
+* **Phase 3 (In Progress)** - OpenShift Virtualization will transition to swap operator.
+  OpenShift will [allow](#swap-ga-for-cnv-users-only) SWAP to be configured only if the swap operator is installed on the cluster.
+* **Phase 4 (TBD)** - OpenShift will allow every user to opt-in for swap. Swap operator will become stand-alone operator deployed by OLM.
+
+### Swap Operator
+
+The operator will be responsible for the following tasks:
+* Automatic deployment of WASP agent and its machine-level configuration.
+* Ease the WASP agent upgrade process i.e. bugfix delivery via z-stream.
+* Limit the usage of swap only for OCP Virt users in phase 3.
+
+MCO will make sure that only swap operator will be eligible to configure swap via the KubeletConfig.
+This will enable Openshift to opt-in for the NodeSwap FG in its earlier stage before it will be GAed in upstream.
+Operator API and its graduation from CNV to OLM will be discussed in a dedicated proposal.
+
+### Workflow Description
+
+**cluster administrator** (Cluster Admin) is a human user responsible
+for deploying and administering a cluster.
+
+**virtual machine owner** (VM Owner) is a human user operating a
+virtual machine in a cluster.
+
+#### Workflow: Configuring higher workload density
+
+**Phases 1-2**
+1. The cluster admin is deploying a bare-metal OpenShift cluster
+2. The cluster admin is deploying the OpenShift Virtualization Operator
+3. The cluster admin is enabling swap on Openshift according to the
+   [documentation](https://github.com/openshift-virtualization/wasp-agent/blob/main/docs/configuration.md)
+4. The cluster admin is configuring OpenShift Virtualization for VM memory overcommit via:
+
+   a. the OpenShift Virtualization Console "Settings" page 
+   b. [or `HCO` API](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/cluster-configuration.md#configure-higher-workload-density)
+
+**Phases 3-4**
+1. The cluster admin is deploying a bare-metal OpenShift cluster.
+2. The cluster admin is deploying the OpenShift Virtualization Operator. 
+3. The cluster admin is configuring the swap operator.
```

Comment:
probably deploying the swap operator too or are we bundling with virt operator to start?

Reply:
You are correct, I've divided between phase 3 and phase 4, as it appears to be different steps.